### PR TITLE
chore: replace lerna with nx, remove unused modules

### LINKED ISSUE #202: chore: remove extends from nx.json
Based on the latest changes to Lerna and NX and a discussion [here](https://github.com/lerna/lerna/issues/3566), it appears that `"extends": "nx/presets/npm.json",` can (and probably should) be removed from the `nx.json` file.
The relevant comment is available [here](https://github.com/lerna/lerna/issues/3566#issuecomment-1463760837).

### LINKED ISSUE #209: refactor: remove unused modules
# Description

1. The conversion from HTML to PDF got started a long time ago and never got completed and tested.
2. The generation of MJML template started more than 2 years ago, but never got completed and used, if we ever do it, t should be with a separate packages to avoid loading too many dependencies in generate-html.

I suggest to remove those modules which would allow us to remove quite a lot of dependencies.

### LINKED ISSUE #210: chore: schema-tools should be migrated to a full NX monorepo
# Description

Lerna should be removed and schema-tools should be migrated to a full NX monorepo

### PR DESCRIPTION
**Pull request by eamon0989**
# Description

We have been facing issues from having a mix of Lerna and Nx, this PR replaces Lerna completely with Nx.
Removes mjml logic and html to pdf.
Removes unneeded `extends` section from `nx.json`.
We previously used parcel to generate the `generateContent.js` files, but it doesn't have a plugin that works with NX, so we switched to @nx/esbuild.
esbuild by default emits esm files, which are not compatible with jest, so I've set the output to be a commonjs file. The code has been updated to expect a `.cjs` file.
We previously had a second "browser" build for 2 packages, this has been removed and the karma tests have also been removed.

Fixes #210 
Fixes #209
Fixes #202  

## Type of change

- [x] Refactor (a code change that neither fixes a bug nor adds a feature)
- [x] Dependencies update

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -56,66 +53,96 @@
...
+    "@types/lodash.get": "^4.4.6",
+    "@types/lodash.groupby": "^4.6.6",
+    "@types/lodash.merge": "^4.6.6",
+    "@types/lodash.set": "^4.3.6",
+    "@types/node": "~20.2.5",
+    "@types/pdfmake": "^0.2.0",
     "@typescript-eslint/eslint-plugin": "^5.12.0",
     "@typescript-eslint/parser": "^5.12.0",
     "audit-ci": "^5.1.2",
     "browserfs": "^1.4.3",
```
Comment (getlarge) [inline package.json:114]: `browserfs` is probably not needed anymore as we removed browser support

File: package.json [hunk 2]
```diff
@@ -56,66 +53,96 @@
...
-    "karma-cli": "^2.0.0",
-    "karma-jasmine": "^5.1.0",
-    "karma-typescript": "^5.5.3",
-    "karma-webpack": "^5.0.0",
+    "jest": "^29.4.1",
+    "jest-environment-node": "^29.4.1",
     "lato-font": "^3.0.0",
-    "lerna": "^5.5.2",
     "lint-staged": "^12.3.4",
     "node-polyfill-webpack-plugin": "^1.1.4",
```
Comment (getlarge) [inline package.json:134]: "node-polyfill-webpack-plugin": not needed either

File: packages/certificate-model/project.json [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+      }
+    },
+    "test": {
+      "executor": "@nx/jest:jest",
+      "outputs": ["{workspaceRoot}/coverage/{projectRoot}"],
+      "options": {
+        "jestConfig": "packages/certificate-model/jest.config.ts",
+        "passWithNoTests": true,
+        "codeCoverage": true,
+        "coverageReporters": ["json", "lcov"]
```
Comment (getlarge) [inline packages/certificate-model/project.json:36]: could you move all the coverage options in the same place (`coverageDirectory` and `coverageThreshold` ) ?
Same for all packages.

Also we don’t “pass with no tests” :)

File: packages/certificate-model/project.json [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+      "options": {
+        "outputPath": "dist/packages/certificate-model",
+        "tsConfig": "packages/certificate-model/tsconfig.lib.json",
+        "packageJson": "packages/certificate-model/package.json",
+        "main": "packages/certificate-model/src/index.ts",
+        "assets": ["packages/certificate-model/*.md"]
+      }
+    },
+    "publish": {
+      "command": "node tools/scripts/publish.mjs certificate-model {args.ver} {args.tag}",
```
Comment (getlarge) [inline packages/certificate-model/project.json:19]: this default publish command is not really ergonomic, could replace it with this [plugin](https://github.com/bikecoders/ngx-deploy-npm) instead ?
you can then remove the `publish.mjs` script.

File: packages/certificate-model/package.json [hunk 5]
```diff
@@ -19,43 +19,17 @@
...
-  "files": [
-    "dist"
-  ],
+  "main": "src/index.js",
+  "types": "src/index.d.ts",
   "engines": {
     "node": ">=12",
     "npm": ">=6"
   },
   "scripts": {
```
Comment (getlarge) [inline packages/certificate-model/package.json:28]: the command in scripts should be moved as targets in project.json except for `prepublishOnly` that can be removed.
In the end the package.json will only be used when installing the lib from npm.

File: packages/certificate-summary/package.json [hunk 6]
```diff
@@ -30,22 +23,11 @@
...
     "cleanup": "rimraf -rf dist coverage",
-    "compile:node": "tsc -p tsconfig.build.json",
-    "compile": "npm run compile:node",
     "lint": "eslint -c ../../.eslintrc.js \"{src,apps,libs,test}/**/*.ts\" --fix",
     "prepublishOnly": "npm run build",
-    "test": "jest",
-    "test:cov": "rimraf -rf ./coverage && jest --ci --coverage",
-    "test:staged": "jest --runInBand --onlyChanged",
     "scan:deps": "audit-ci --config ../../audit-ci.json",
     "example": "npm run build && node example.js"
```
Comment (getlarge) [inline packages/certificate-summary/package.json:31]: same as certificate-model but lint already as its own target.

File: packages/generate-coa-pdf-template/project.json [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+        },
+        "production": {
+          "minify": true
+        }
+      }
+    },
+    "publish": {
+      "command": "node tools/scripts/publish.mjs generate-coa-pdf-template {args.ver} {args.tag}",
+      "dependsOn": ["build"]
+    },
```
Comment (getlarge) [inline packages/generate-coa-pdf-template/project.json:32]: publish is not needed for this project, `generate-en10168-pdf-template` and `generate-pdf-template-helpers` as well

File: packages/certificate-model/package.json [hunk 8]
```diff
@@ -19,43 +19,17 @@
...
-    "ajv": "8.12.0",
-    "ajv-formats": "2.1.1",
-    "lodash.clonedeepwith": "^4.5.0",
-    "lodash.merge": "^4.6.2"
-  },
-  "devDependencies": {
-    "@types/lodash.clonedeepwith": "^4.5.6",
-    "@types/lodash.merge": "^4.6.6"
-  },
   "gitHead": "5526910fd5974d09e80b5ba6324043b4409dd179"
```
Comment (getlarge) [inline packages/certificate-model/package.json:34]: gitHead can be removed from all package.json,it was used when publishing through lerna.

**Review by getlarge — CHANGES_REQUESTED** [review]
That was lot of work!
There are still few steps to finish the accomplishment as described in this review.
Also we loose some other benefits by moving away from lerna:
- packages versioning (version bump based on commits convention, change log generation…)
- github release

This can be done using this [Nx plugin](https://github.com/jscutlery/semver) and using the **independent** mode. The doc also indicates how to release on GitHub using the `gh` CLI.

File: packages/certificate-model/project.json (hunk 3 above)
Comment (eamon0989) [reply packages/certificate-model/project.json:36]: > Also we don’t “pass with no tests” :)

😂

File: packages/certificate-model/project.json (hunk 4 above)
Comment (eamon0989) [reply packages/certificate-model/project.json]: Done. `lerna version` has obviously been removed, should something like https://github.com/jscutlery/semver be added to take care of versioning? I see you used it in your `nx-heroku` project.

Comment (getlarge) [reply packages/certificate-model/project.json]: yes, that’s the same `Nx Plugin` i mentioned in the comment above.

File: utils/create-schema-interfaces-interactive.ts [hunk 9]
```diff
@@ -1,9 +1,12 @@
...
-const fs = require('fs');
+/* eslint-disable no-console */
+import fs from 'fs';
+import path from 'path';
+import { stdin as input, stdout as output } from 'process';
+import readline from 'readline';
+
+// TODO: fix module not found error when called from packages/generate-coa-pdf-template/project.json
+// could this be moved to packages/generate-coa-pdf-template/utils?
+import { generate } from '@s1seven/schema-tools-generate-interfaces';
```
Comment (getlarge) [inline utils/create-schema-interfaces-interactive.ts:13]: I did not encounter an issue with the import, but the real problem is that the default Nx executor (nx:run-commands) does not support interactive CLI.
So this should be run directly via ts-node.

**Review by getlarge — APPROVED** [review]

**Pull request merged**